File: modules/pdf-pages/android/src/main/java/expo/modules/pdfpages/PdfPagesModule.kt
package expo.modules.pdfpages

import android.graphics.Bitmap
import android.graphics.Color
import android.graphics.pdf.PdfRenderer
import android.net.Uri
import android.os.ParcelFileDescriptor
import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.io.File
import java.io.FileOutputStream

class PdfUnreadableException(cause: String) :
  CodedException("ERR_PDF_UNREADABLE", "Could not read the PDF: $cause", null)

/**
 * PDF → PNG pages, so the score viewer only ever draws images.
 *
 * Mirrors the iOS module: same file names (`1.png`, `2.png`, …), same page
 * order, same "scale the longest edge to maxWidth" rule. Rendering happens
 * once at import; the viewer never touches a PDF.
 */
class PdfPagesModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("PdfPages")

    AsyncFunction("render") { uri: String, outDir: String, maxWidth: Double ->
      render(uri, outDir, maxWidth)
    }
  }

  private fun toFile(s: String): File =
    if (s.startsWith("file://")) File(Uri.parse(s).path ?: s.removePrefix("file://")) else File(s)

  private fun render(uri: String, outDir: String, maxWidth: Double): List<Map<String, Any>> {
    val src = toFile(uri)
    if (!src.exists()) throw PdfUnreadableException("no file at $uri")
    val dir = toFile(outDir).also { it.mkdirs() }

    val fd = try {
      ParcelFileDescriptor.open(src, ParcelFileDescriptor.MODE_READ_ONLY)
    } catch (e: Exception) {
      throw PdfUnreadableException(e.message ?: "open failed")
    }
    // A PdfRenderer holds the descriptor, so both close together or the fd leaks.
    fd.use { descriptor ->
      val renderer = try {
        PdfRenderer(descriptor)
      } catch (e: Exception) {
        throw PdfUnreadableException(e.message ?: "not a PDF, or password protected")
      }
      renderer.use { pdf ->
        return (0 until pdf.pageCount).map { i ->
          pdf.openPage(i).use { page ->
            // scale the longest edge down to maxWidth; never scale a small page up
            val scale = minOf(1.0, maxWidth / maxOf(page.width, page.height).toDouble())
            val w = maxOf(1, (page.width * scale).toInt())
            val h = maxOf(1, (page.height * scale).toInt())
            val bmp = Bitmap.createBitmap(w, h, Bitmap.Config.ARGB_8888)
            // PdfRenderer draws onto transparency; sheet music on a transparent
            // background disappears in dark mode, so paint the paper first.
            bmp.eraseColor(Color.WHITE)
            page.render(bmp, null, null, PdfRenderer.Page.RENDER_MODE_FOR_DISPLAY)
            val out = File(dir, "${i + 1}.png")
            FileOutputStream(out).use { bmp.compress(Bitmap.CompressFormat.PNG, 100, it) }
            bmp.recycle()
            mapOf("uri" to Uri.fromFile(out).toString(), "width" to w, "height" to h)
          }
        }
      }
    }
  }
}

File: modules/pdf-pages/ios/PdfPagesModule.swift
import ExpoModulesCore
import PDFKit
import UIKit

/**
 * PDF → PNG pages, so the score viewer only ever draws images.
 *
 * Mirrors the Android module exactly: same file names (`1.png`, `2.png`, …),
 * same page order, same "scale the longest edge to maxWidth" rule. Rendering
 * happens once at import; the viewer never touches a PDF.
 *
 * NOTE: written without a device to verify against — this project has no Apple
 * developer account yet. Test before trusting it.
 */
public class PdfPagesModule: Module {
  public func definition() -> ModuleDefinition {
    Name("PdfPages")

    AsyncFunction("render") { (uri: String, outDir: String, maxWidth: Double) -> [[String: Any]] in
      try self.render(uri: uri, outDir: outDir, maxWidth: maxWidth)
    }
  }

  private func toURL(_ s: String) -> URL {
    s.hasPrefix("file://") ? (URL(string: s) ?? URL(fileURLWithPath: s)) : URL(fileURLWithPath: s)
  }

  private func render(uri: String, outDir: String, maxWidth: Double) throws -> [[String: Any]] {
    let src = toURL(uri)
    guard let doc = PDFDocument(url: src) else {
      throw Exception(name: "ERR_PDF_UNREADABLE", description: "Could not read the PDF at \(uri)")
    }
    let dir = toURL(outDir)
    try FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)

    var pages: [[String: Any]] = []
    for i in 0..<doc.pageCount {
      guard let page = doc.page(at: i) else { continue }
      let box = page.bounds(for: .mediaBox)
      // scale the longest edge down to maxWidth; never scale a small page up
      let scale = min(1.0, maxWidth / max(box.width, box.height))
      let size = CGSize(width: max(1, box.width * scale), height: max(1, box.height * scale))

      let format = UIGraphicsImageRendererFormat()
      format.scale = 1 // size is already in pixels, not points
      format.opaque = true
      let image = UIGraphicsImageRenderer(size: size, format: format).image { ctx in
        // sheet music on a transparent background disappears in dark mode
        UIColor.white.setFill()
        ctx.fill(CGRect(origin: .zero, size: size))
        // PDF space has its origin bottom-left; UIKit's is top-left
        ctx.cgContext.translateBy(x: 0, y: size.height)
        ctx.cgContext.scaleBy(x: scale, y: -scale)
        ctx.cgContext.translateBy(x: -box.origin.x, y: -box.origin.y)
        page.draw(with: .mediaBox, to: ctx.cgContext)
      }
      guard let png = image.pngData() else { continue }
      let out = dir.appendingPathComponent("\(i + 1).png")
      try png.write(to: out, options: .atomic)
      pages.append(["uri": out.absoluteString, "width": Int(size.width), "height": Int(size.height)])
    }
    return pages
  }
}
